neuro_san/internals/run_context/langchain/base_tool_factory.py

```diff
@@ -10,106 +10,232 @@
+        :return: - Instantiated tool if "class" of tool_name points to a BaseTool class
+                 - A list of tools if "class of "tool_name points to a BaseToolkit class.
+        """
+        empty: Dict[str, Any] = {}
+
+        tool_info: Dict[str, Any] = self.base_tool_infos.get(tool_name)
+        if not tool_info:
+            raise ValueError(f"Tool '{tool_name}' is not defined in {self.base_tool_info_file}.")
+
+        if not isinstance(tool_info, Dict):
+            raise ValueError(f"The value for the {tool_name} key must be a dictionary.")
+
+        if "class" not in tool_info:
+            raise ValueError(
+                "Missing required key: 'class'.\n"
+                "Each tool configuration must include a 'class' key specifying the fully qualified class name.\n"
+                "Example: 'class': 'langchain_community.tools.bing_search.BingSearchResults'."
             )
-            request_tools = request_toolkit.get_tools()
-            mapping = {
-                "requests_get": request_tools[0],
-                "requests_post": request_tools[1],
-                "requests_patch": request_tools[2],
-                "requests_put": request_tools[3],
-                "requests_delete": request_tools[4],
-            }
-            return mapping[tool_name]
-
-        raise ValueError(
-            f"""Tool '{tool_name}' is not supported.
-            Ensure that you have provided a valid tool name from the supported list:
-            'bing_search', 'tavily_search', 'requests_get', 'requests_post', 'requests_patch', 'requests_put',
-            'requests_delete'.
-
-            If you are trying to use a different tool, you may need to implement support for it within
-            the PrebuiltBaseToolFactory class. To do this, extend the `get_agent_tool` method
-            by adding a corresponding condition for your new tool.
-
-            Refer to the class documentation for more details on extending this functionality."""
-        )
+
+        # Instantiate the main tool or toolkit class
+        tool_class: Type[Any] = self._resolve_class(tool_info.get("class"))
+        # Recursively resolve arguments (including wrapper dependencies)
+        resolved_args: Dict[str, Any] = self._resolve_args(tool_info.get("args", empty))
+        # Merge with user arguments where user_args get the priority
+        final_args: Dict[str, Any] = self.overlayer.overlay(resolved_args, user_args) if user_args else resolved_args
+
+        # Use the "from_{tool_name}_api_wrapper" method if available, otherwise the constructor
+        callable_obj: Union[Type[BaseTool], Type[BaseToolkit], Callable[..., Any]] = \
+            self._get_from_api_wrapper_method(tool_class) or tool_class
+
+        # Validate and instantiate
+        self._check_invalid_args(callable_obj, final_args)
+        # Instance can be a BaseTool or a BaseToolkit
+        instance: Union[BaseTool, BaseToolkit] = callable_obj(**final_args)
```

Comment:
Similar deal with catching exceptions when this fails as for nested_class() below.
How has this been resolved?